File: 9-Stack/programs/ruby/program6.rb
# Program for evaluation of postfix expression

class Stack
  def initialize
    @stack = []
  end

  def push(x)
    @stack.push(x)
  end

  def pop
    if @stack.empty?
      puts "Stack is Empty"
      return -1
    else
      return @stack.pop
    end
  end

  def display
    puts @stack.join(' ')
  end
end

def is_operand(x)
  !['+', '-', '*', '/'].include?(x)
end

def eval_postfix(postfix)
  stack = Stack.new
  postfix.each_char do |char|
    if is_operand(char)
      stack.push(char.to_i)
    else
      x2 = stack.pop
      x1 = stack.pop
      result = case char
               when '+' then x1 + x2
               when '-' then x1 - x2
               when '*' then x1 * x2
               when '/' then x1 / x2
               end
      stack.push(result)
    end
  end
  stack.pop
end

postfix = "234*+82/-"
puts "Result is #{eval_postfix(postfix)}"

File: 9-Stack/programs/cpp/program6.cpp
// Program for evaluation of postfix expression
#include<bits/stdc++.h>
using namespace std;

struct Node{
  char data;
  struct Node *next;
}*top = NULL;

void push(char x){
  struct Node *t;
  t = new Node;
  if(t == NULL){
    cout << "Stack is full" << endl;
  }else{
    t->data = x;
    t->next = top;
    top = t;
  }
}

char pop(){
  char x = -1;
  if(top == NULL){
    cout << "Stack is Empty" << endl;
  }else{
    struct Node *t = top;
    top = top->next;
    x = t->data;
    delete t;
  }
  return x;
}

void Display(){
  struct Node *p = top;
  while(p != NULL){
    cout << p->data << " ";
    p = p->next;
  }
  cout << endl;
}

int pre(char x){
  if(x == '+' || x == '-'){
    return 1;
  }else if(x == '*' || x == '/'){
    return 2;
  }
  return 0;
}

int isOperand(char x){
  if(x == '+' || x == '-' || x == '*' || x == '/'){
    return 0;
  }else{
    return 1;
  }
}

int eval(char *postfix){
  int i = 0;
  int x1, x2, r;
  for(i = 0; postfix[i] != '\0'; i++){
    if(isOperand(postfix[i])){
      push(postfix[i] - '0');
    }else{
      x2 = pop();
      x1 = pop();
      switch(postfix[i]){
        case '+': r = x1 + x2; break;
        case '-': r = x1 - x2; break;
        case '*': r = x1 * x2; break;
        case '/': r = x1 / x2; break;
      }
      push(r);
    }
  }
  return top->data;
}

int main(){
  char *postfix = "234*+82/-";
  cout << "Result is " << eval(postfix) << endl;
  return 0;
}
// Output:
// Result is 2
